torch/_functorch/autograd_function.py

```diff
@@ -0,0 +1,196 @@
+import torch
+from torch._ops import PyOperator
+from torch._C._functorch import TransformType
+from torch._functorch.utils import enable_autograd_function
+from torch.autograd.function import _SingleLevelFunction
+import torch.utils._pytree as pytree
+from torch._C._functorch import (
+    _wrap_for_grad,
+    _unwrap_for_grad,
+)
+
+# autograd.Function technically runs before the regular PyTorch dispatcher.
+# This is how features like autocast and torch_dispatch (e.g. PythonTLSSnapshot)
+# work with it. One day we might decide to change this, but until then,
+# we need to give the illusion that autograd.Function runs before those things.
+#
+# We do this by using creating a custom PyOperator that only functorch
+# dispatching specially.
+class CustomFunctionPyOperator(PyOperator):
+    def __init__(self):
+        super().__init__('custom_function_call')
+
+    def __call__(self, *args, **kwargs):
+        # When custom_function_call is done dispatching through functorch,
+        # it should just invoke the autograd.Function. This is consistent
+        # with the autograd.Function behavior of being invoked before the
+        # PyTorch dispatcher.
+        #
+        # This will lead us into trouble later down the line, but this is
+        # pre-existing: one is unable to trace autograd.Function calls with
+        # ProxyTensor/AOTDispatcher.
+        if torch._C._are_functorch_transforms_active():
+            return super().__call__(*args, **kwargs)
+        autograd_function = args[0]
+        return autograd_function.apply(*args[1:], **kwargs)
+
+
+# "custom_function_call"
```

Comment:
Naming question: I refer to autograd.Function as "AutogradFunction" sometimes, and sometimes as "CustomFunction". What do you guys want me to call it?

Reply:
Should the instance of CustomFunctionPyOperator be private? I see it being called as part of "apply', but is there a case when someone would someone call it directly?

Reply:
Everything inside `torch/_functorch/autograd_function.py` is private, including custom_function_call. Users should not be calling it directly.

Developers should interact with custom_function_call directly, instead of CustomFunctionPyOperator if that is the question. custom_function_call is an instance of a PyOperator, CustomFunctionPyOperator is just a class.